nengo_spa/compiler/parsing.py

```diff
@@ -0,0 +1,319 @@
+"""
+
+from nengo_spa.compiler import tokens
+
+
+class RuleMismatchError(Exception):
+    """Raised when trying to apply a rule to tokens that do not match the rule.
+    """
+    def __init__(self, rule, expected, found):
+        super(RuleMismatchError, self).__init__(
+            "Could not apply rule '{}'. Expected '{}', but found '{}'.".format(
+                rule, expected, found))
+
+
+class MatchToken(object):
+    """Class to match `TokenInfo` objects.
+
+    This class will be equal to any `TokenInfo` object that matches all given
+    arguments with its attributes. It is allowed to have further arguments that
+    are not required to match.
+
+    Arguments
+    ---------
+    kwargs : dict
+        Definition of attributes that need to match.
+    """
+
+    arg2idx = {arg: idx for idx, arg in enumerate(
+        ('type', 'string', 'start', 'end', 'line'))}
+
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+
+    def __eq__(self, other):
+        return isinstance(other, tokens.TokenInfo) and all(
+            self._match_arg(other, arg) for arg in self.kwargs)
+
+    def _match_arg(self, other, arg):
+        if hasattr(other, arg):
+            return getattr(other, arg) == self.kwargs[arg]
+        else:
+            return other[self.arg2idx[arg]] == self.kwargs[arg]
+
+    @classmethod
+    def from_token(cls, token, match_on=None):
+        """Creates an instance from given `TokenInfo` object.
+
+        By default *type* and *string* need to match.
+
+        Arguments
+        ---------
+        token : TokenInfo
+            Token to create the instance from.
+        match_on : sequence, optional
+            Sequence of attributes that need to be matched.
+        """
+        if match_on is None:
+            match_on = ['type', 'string']
+        return cls(**{
+            name: getattr(token, name)
+            if hasattr(token, name) else token[cls.arg2idx[name]]
+            for name in match_on})
+
+    def __repr__(self):
+        return 'MatchToken({})'.format(
+            ', '.join('{}={!r}'.format(k, v) for k, v in self.kwargs.items()))
+
+
+class Rule(object):
+    """Abstract base class for grammar rules.
+
+    Rules can be chained with the addition operator ``+`` and alternatives can
+    be given with the or operator ``|``. Note that the first alternative that
+    matches the upcoming tokens will be applied without any form of
+    backtracking. Thus, the order when listing alternatives may matter.
+    """
+
+    def accept(self, tokens):
+        """Tests whether the upcoming tokens are accepted by the rule.
+
+        Parameters
+        ----------
+        tokens : nengo.internal.generators.Peekable
+            Peekable generator providing the upcoming tokens.
+
+        Returns
+        -------
+        bool
+        """
+        raise NotImplementedError()
+
+    def read(self, tokens):
+        """Reads the tokens excepted by the rule and returns derived objects.
```

Comment:
Is this supposed to be "accepted" or "excepted"?

Reply:
I think it can be either. The rule defines certain tokens it expects which are the same tokens that need to be there for the rule to accept the tokens. However, the `accept` method only looks one (maybe two) tokens ahead and doesn't check all expected tokens.